relay-system/src/service.rs

```diff
@@ -44,45 +116,313 @@ impl fmt::Display for SendError {
+pub struct Addr<I: Interface> {
+    tx: mpsc::UnboundedSender<I>,
+}
+
+impl<I: Interface> fmt::Debug for Addr<I> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Addr")
+            .field("open", &!self.tx.is_closed())
+            .finish()
+    }
+}
+
+// Manually derive `Clone` since we do not require `I: Clone` and the Clone derive adds this
 // constraint.
-impl<S: Service> Clone for Addr<S> {
+impl<I: Interface> Clone for Addr<I> {
     fn clone(&self) -> Self {
         Self {
             tx: self.tx.clone(),
         }
     }
 }
 
-impl<S: Service> Addr<S> {
-    /// Sends an asynchronous message to the service and waits for the response.
+impl<I: Interface> Addr<I> {
+    /// Sends a message to the service and returns the response.
     ///
-    /// The result of the message does not have to be awaited. The message will be delivered and
-    /// handled regardless. The communication channel with the service is unbounded, so backlogs
-    /// could occur when sending too many messages.
+    /// Depending on the message's response behavior, this either returns a future resolving to the
+    /// return value, or does not return anything for fire-and-forget messages. The communication
+    /// channel with the service is unbounded, so backlogs could occur when sending too many
+    /// messages.
     ///
-    /// Sending the message can fail with `Err(SendError)` if the service has shut down.
-    // Note: this is written as returning `impl Future` instead of `async fn` in order not
-    // to capture the lifetime of `&self` in the returned future.
-    pub fn send<M>(&self, message: M) -> impl Future<Output = Result<M::Response, SendError>>
+    /// Sending asynchronous messages can fail with `Err(SendError)` if the service has shut down.
+    /// The result of asynchronous messages does not have to be awaited. The message will be
+    /// delivered and handled regardless:
+    pub fn send<M>(&self, message: M) -> <I::Response as MessageResponse>::Output
     where
-        M: ServiceMessage<S>,
+        I: FromMessage<M>,
     {
-        let (messages, response_rx) = message.into_messages();
-        let res = self.tx.send(messages);
-        async move {
-            res.map_err(|_| SendError)?;
-            response_rx.await.map_err(|_| SendError)
-        }
+        let (tx, rx) = I::Response::channel();
+        self.tx.send(I::from_message(message, tx)).ok(); // it's ok to drop, the response will fail
+        rx
+    }
+}
+
+/// Inbound channel for messages sent through an [`Addr`].
+///
+/// This channel is meant to be polled in a [`Service`].
+///
+/// Instances are created automatically when [running](Service::run) a service, or can be
+/// created through [`channel`]. The channel closes when all associated [`Addr`]s are dropped.
+pub type Receiver<I> = mpsc::UnboundedReceiver<I>;
+
+/// Creates an unbounded channel for communicating with a [`Service`] without backpressure.
```

Comment:
```suggestion
/// Creates an unbounded channel for communicating with a [`Service`].
```

This is just a nit, it's totally true as written.  But backpressure is more complicated than simply using a bounded channel.  Technical writing guidelines tends to encourage removing superfluous bits though ;)